### DISCUSSION TITLE
Merger with BLE and Serial1 - CC events not forwarded

### DISCUSSION THREAD

**Discussion by agentff6600** [FortySevenEffects/arduino_midi_library#283] [Q&A]
Hi!

I am working on a merger that repeats messages between MIDI I/O via serial1 and BLE MIDI.

Any message gets through from MIDI in to BLE MIDI
From BLE MIDI to MIDI out I only get note events.

Any tips, pointers, remarks? I used the example "Dual Merger" as base.

```
#include <BLEMIDI_Transport.h>
#include <hardware/BLEMIDI_ArduinoBLE.h>


BLEMIDI_CREATE_INSTANCE("Soma_MIDI_Interface", MIDI);
MIDI_CREATE_INSTANCE(HardwareSerial, Serial1, midiB);

unsigned long tClock = millis();
bool isConnected = false;

using namespace MIDI_NAMESPACE;

void setup()
{
    
    MIDI.begin(MIDI_CHANNEL_OMNI);
    midiB.begin(MIDI_CHANNEL_OMNI);

  pinMode(LED_BUILTIN, OUTPUT);

  digitalWrite(LED_BUILTIN, LOW);
  
  BLEMIDI.setHandleConnected([]() {

    isConnected = true;
    digitalWrite(LED_BUILTIN, HIGH);

    if ((millis() - tClock) > 20)
  {
     tClock = millis();
     MIDI.sendRealTime(MidiType::Clock);
  }

  });

  BLEMIDI.setHandleDisconnected([]() {

    isConnected = false;
    digitalWrite(LED_BUILTIN, LOW);

  });

}

void loop()
{
    if (MIDI.read())
    {
        midiB.send(MIDI.getType(),
                   MIDI.getData1(),
                   MIDI.getData2(),
                   MIDI.getChannel());
    }

    if (midiB.read())
    {

        MIDI.send(midiB.getType(),
                   midiB.getData1(),
                   midiB.getData2(),
                   midiB.getChannel());
    }
}
```

**Comment by franky47** [comment]
cc @lathoub, any ideas?

**Comment by lathoub** [comment]
hmm, good question. Do you get all traffic when using **just** BLE @agentff6600 ? (so removing all serial MIDI from the example)

For these sort of use-case (merger), I add a callback to:
https://github.com/FortySevenEffects/arduino_midi_library/blob/2d64cc3c2ff85bbee654a7054e36c59694d8d8e4/src/MIDI.h#L186

That way, you are sure to get the entire, correct, MIDI message. 

Also, add an [error handler](https://github.com/FortySevenEffects/arduino_midi_library/blob/master/examples/ErrorCallback/ErrorCallback.ino) to make sure no errors are reported.

**Reply by agentff6600** [reply]
Thank you for  the reply! The BLE works on its own. I am using a MKR1010.

I am although walking around in the dark when it comes to the Handler that you suggested.

I tried the following code (based on the issue #159):

```
#include <BLEMIDI_Transport.h>
#include <hardware/BLEMIDI_ArduinoBLE.h>
#include <MIDI.h>


BLEMIDI_CREATE_INSTANCE("Soma_MIDI_Interface", MIDI);
using namespace MIDI_NAMESPACE;

typedef Message<MIDI_NAMESPACE::DefaultSettings::SysExMaxSize> MidiMessage;

MIDI_CREATE_INSTANCE(HardwareSerial, Serial1, midiB);

unsigned long tClock = millis();
bool isConnected = false;



void setup()
{

    // Initiate MIDI communications, listen to all channels
    MIDI.begin(MIDI_CHANNEL_OMNI);
    midiB.begin(MIDI_CHANNEL_OMNI);

  pinMode(LED_BUILTIN, OUTPUT);

  digitalWrite(LED_BUILTIN, LOW);
  
  BLEMIDI.setHandleConnected([]() {

    isConnected = true;
    digitalWrite(LED_BUILTIN, HIGH);

    if ((millis() - tClock) > 20)
  {
     tClock = millis();
     MIDI.sendRealTime(MidiType::Clock);
  }

  });

  BLEMIDI.setHandleDisconnected([]() {

    isConnected = false;
    digitalWrite(LED_BUILTIN, LOW);

  });

    MIDI.setHandleMessage(OnBLEMessage);
    midiB.setHandleMessage(OnSerialMessage);

}

void loop()
{
    MIDI.read();
    midiB.read();
}

void OnBLEMessage(const MidiMessage& message)
{
  midiB.send(message);
}

void OnSerialMessage(const MidiMessage& message)
{
  MIDI.send(message);
}

```

But with this it does not work at all.

**Reply by lathoub** [reply]
1) connecting code:

```cpp
  BLEMIDI.setHandleConnected([]() {
    isConnected = true;
    digitalWrite(LED_BUILTIN, HIGH);
    if ((millis() - tClock) > 20)
    {
       tClock = millis();
       MIDI.sendRealTime(MidiType::Clock);
    }
  });
```

not sure if I understand, it seems this is a _leftover_ from `loop`

2) if I replace the above with 

```cpp
  BLEbleMIDI.setHandleConnected([]() {
    digitalWrite(LED_BUILTIN, HIGH);
  });
```
the device does not want to stay connected anymore.

This discussion does not belong in this repo, and should be transferred to [BLEMIDI](https://github.com/lathoub/Arduino-BLE-MIDI) as an _issue_.

**Reply by lathoub** [reply]
> the device does not want to stay connected anymore.

This happens when not setting an initial value in the characteristic:

https://github.com/lathoub/Arduino-BLE-MIDI/blob/edd2f21a399c09a4daaf095ea07c8e8d9669e473/src/hardware/BLEMIDI_ArduinoBLE.h#L209

Fixed. Please try again with the [CustomSettings branch op Arduino-BLE-MIDI](https://github.com/lathoub/Arduino-BLE-MIDI/tree/CustomSettings)

```cpp
#include <BLEMIDI_Transport.h>
#include <hardware/BLEMIDI_ArduinoBLE.h>
using namespace MIDI_NAMESPACE;

typedef Message<MIDI_NAMESPACE::DefaultSettings::SysExMaxSize> MidiMessage;

BLEMIDI_CREATE_INSTANCE("Soma_MIDI_Interface", MIDI);
MIDI_CREATE_INSTANCE(HardwareSerial, Serial1, midiB);

void setup()
{
  Serial.begin(115200);
  while (!Serial);
  Serial.println("Ready");

  MIDI.begin(MIDI_CHANNEL_OMNI);
  midiB.begin(MIDI_CHANNEL_OMNI);

  pinMode(LED_BUILTIN, OUTPUT);
  digitalWrite(LED_BUILTIN, LOW);

  BLEMIDI.setHandleConnected([]() {
    Serial.println("BLE Connected");
    digitalWrite(LED_BUILTIN, HIGH);
  });

  BLEMIDI.setHandleDisconnected([]() {
    digitalWrite(LED_BUILTIN, LOW);
    Serial.println("BLE Disconnected");
  });

  MIDI.setHandleMessage(OnBLEMessage);
  midiB.setHandleMessage(OnSerialMessage);
}

void loop()
{
  MIDI.read();
  midiB.read();
}

void OnBLEMessage(const MidiMessage& message)
{
  Serial.println("BLE-MIDI Message received");
  midiB.send(message);
}

void OnSerialMessage(const MidiMessage& message)
{
  Serial.println("Serial MIDI Message received");
  MIDI.send(message);
}
```

**Reply by agentff6600** [reply]
Thank you so much for this.

BLE connects and the serial terminal spits out "message received" from both handlers.
Although the message doesn't show up on my Midi monitor connected via BLE nor does hardware connected via serial receive any messages from BLE.

The hardware serial MIDI IN and OUT works.  All messages are passed through.

**Reply by lathoub** [reply]
Progress! But not sure if I understand what you are saying.

So only `MIDI.send(message);` in the above example does **NOT** work? `midiB.send(message);` works fine?
